driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -137,6 +137,9 @@ public String getLoggedKeyspace() {
 
   @Override
   public ResultSetFuture executeAsync(final Statement statement) {
+    if (cluster.isClosed()) {
```

Comment:
I think this isn't needed, since this ultimately calls the `execute` method you also have logic for further below.  It's better to just rely on that since it will fail the future instead of the calling thread which is nicer for users since then they don't need to handle exceptions in two different places.

Reply:
Hmm it might be required, `isInit` remains true even after the cluster is closed, and in this case we don't want the flow to reach the if block below.

Reply:
I added this specifically for isInit case.

Reply:
If it's still required, we should change it to fail the future, not throw an Exception.